### PR TITLE
Implement a user-friendly refine function making locate more modular

### PR DESCRIPTION
**Pull request by caspervdw**
This continues the work discussed in #257.

`trackpy.refine_com` now returns a `DataFrame` with clearly labeled columns. `locate` is now rebased on this, and as it does not require the to-dataframe conversion any more, this simplifies `locate` a bit. There might be some performance loss in locate, although I expect it is negligible.

For users that want to use only numpy, `trackpy.refine_com_arr` is there, which is basically the old `refine` function.

### CODE REVIEW

**Comment by nkeim** [comment]
I am benchmarking this and will let you know the results.

**Comment by nkeim** [comment]
No performance loss. http://soft-matter.github.io/trackpy-bench/index.html#real_data.DenseFeatureSuite.time_numba

**Comment by danielballan** [comment]
:+1: for this approach

**Comment by caspervdw** [comment]
OK I pushed some fixes so this is good to go.

File: trackpy/refine/center_of_mass.py [hunk 1]
```diff
@@ -53,23 +55,59 @@ def refine_com(raw_image, image, radius, coords, separation=0, max_iterations=10
...
-    if break_thresh is not None:
-        warnings.warn("break_threshold has been deprecated: shift_threshold is"
-                      "the only parameter that determines when to shift the"
-                      "mask.", DeprecationWarning)
     if max_iterations <= 0:
         warnings.warn("max_iterations has to be larger than 0. setting it to 1.")
         max_iterations = 1
-    # ensure that radius is tuple of integers, for direct calls to refine()
+    if raw_image.ndim != coords.shape[1]:
+        raise ValueError("The image has a differnt number of dimensions than "
```
Comment (nkeim) [inline trackpy/refine/center_of_mass.py]: different

File: trackpy/refine/center_of_mass.py [hunk 2]
```diff
@@ -39,7 +41,7 @@ def refine_com(raw_image, image, radius, coords, separation=0, max_iterations=10
         used for final characterization
     image : array (any dimension)
         processed image, used for locating center of mass
-    coord : array
+    coord : array or DataFrame
```
Comment (nkeim) [inline trackpy/refine/center_of_mass.py]: coords

File: trackpy/refine/center_of_mass.py [hunk 3]
```diff
@@ -53,23 +55,59 @@ def refine_com(raw_image, image, radius, coords, separation=0, max_iterations=10
...
+    else:
+        index = None
+
+    radius = validate_tuple(radius, image.ndim)
+
+    if pos_columns is None:
+        pos_columns = default_pos_columns(image.ndim)
+    columns = pos_columns + ['mass']
+    if characterize:
+        isotropic = radius[1:] == radius[:-1]
```
Comment (nkeim) [inline trackpy/refine/center_of_mass.py]: Is this supposed to be a boolean array or a scalar?

File: trackpy/feature.py [hunk 4]
```diff
@@ -355,72 +301,69 @@ def locate(raw_image, diameter, minmass=None, maxsize=None, separation=None,
...
+        condition &= refined_coords['size'] < maxsize
+    if not condition.all():  # apply the filter
+        # making a copy to avoid SettingWithCopyWarning
+        refined_coords = refined_coords.loc[condition].copy()
 
-    if count_qualified == 0:
+    if len(refined_coords) == 0:
         warnings.warn("No maxima survived mass- and size-based filtering. "
                       "Be advised that the mass computation was changed from "
                       "version 0.2.4 to 0.3.0. See the documentation and the "
```
[the remark this answers is not in the document]
Comment (caspervdw) [reply trackpy/feature.py]: That has to go with PR #465, I'll push the change there.

File: trackpy/refine/center_of_mass.py (hunk 3 above)
Comment (caspervdw) [reply trackpy/refine/center_of_mass.py]: Because of the `validate_tuple` above, `radius` will always be a tuple so that `isotropic` will be a scalar.

**Comment by caspervdw** [comment]
@nkeim Happy newyear! I looked at your review and pushed some changes.

**Comment by nkeim** [comment]
@caspervdw Happy new year! This looks good. Looks like a rebase is now required (sorry!).

**Comment by caspervdw** [comment]
Rebased

**Comment by caspervdw** [comment]
@nkeim I restarted Travis and it passed, the issues are solved apparently.

**Pull request merged**